### DISCUSSION TITLE
How do decide if a node is a valid expression

### DISCUSSION THREAD

**Discussion by AckslD** [nvim-treesitter/nvim-treesitter#6258] [Q&A]
Is there some way to decide if a node is a valid expression without having to write complicated captures for all languages? As an example, let's say I want to decide if the current visual selection in `Python` is valid to use as a return value in a function.

Related to:
https://github.com/ThePrimeagen/refactoring.nvim/issues/443#issuecomment-1980137717

**Comment by clason** [comment]
Please define "a valid expression". If you mean "syntactically valid" (to the extent that the parser understands it -- not every parser is complete), then you don't need any queries and can just parse the string and check for `(ERROR)` nodes.

**Reply by AckslD** [reply]
In the case of `python` I would like to know if a treesitter node corresponds to one of [these nodes](https://docs.python.org/3/library/ast.html#expressions). I'm basically wondering if treesitter has a language-agnostic notion of this. Sorry if this is still vague, I'm looking for the correct terminology.

**Reply by clason** [reply]
No. All it does is generate a concrete syntax tree, with nodes specified by the parser (so **VERY** language-specific).

**Reply by clason** [reply]
(Such general tree-sitter questions are better asked upstream, e.g., at https://matrix.to/#/#tree-sitter-chat:matrix.org

**Reply by AckslD** [reply]
Thanks! Do you think it's worth raising this question in the matrix forum as well? Basically I'm curious if someone have some clever idea how to do such a check without having a curated list of node types for each language. But maybe that's the only way

**Reply by amaanq** [reply]
it's language specific; the whole point of queries is to unify language constructs under `captures`, though you'd have to write a query for *every* desired language.
